thrifty-runtime/src/jvmMain/kotlin/com/microsoft/thrifty/transport/THttpTransport.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.microsoft.thrifty.transport
+
+
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InputStream
+import java.net.HttpURLConnection
+import java.net.URL
+
+/**
+ * HTTP implementation of the TTransport interface. Used for working with a
+ * Thrift web services implementation (using for example TServlet).
+ *
+ * THIS IMPLEMENTATION IS NOT THREAD-SAFE !!!
+ *
+ * Based on the official thrift java THttpTransport with the apache client support removed.
+ * Both due to wanting to avoid the additional dependency as well as it being a bit weird to have two
+ * implementations to switch between in the same class.
+ *
+ * Uses HttpURLConnection internally
+ *
+ * Also note that under high load, the HttpURLConnection implementation
+ * may exhaust the open file descriptor limit.
+ *
+ * @see [THRIFT-970](https://issues.apache.org/jira/browse/THRIFT-970)
+ */
+open class THttpTransport(url: String) : Transport {
```

Comment:
Also, a substantive comment - can we please make an `expect class` in `src/commonMain/kotlin/com/microsoft/thrifty/transport` for this?  Then make this one the `actual`?

Generally on JVM I favor using OkHttp over HttpURLConnection - it's much easier to use correctly and scalably.  Is there a reason, aside from the initial implementation using it, to favor HttpURLConnection?

I suppose the fact that it's built in to the JDK is a point in its favor, along with not wanting to bring in extra dependencies.  This is part of why I suggested making a separate module for this guy.

Reply:
What benefit do we get from making it an expect/actual class? Why can't there just be different types of Transports implemented for different variants? This is a genuine question, I have no experience with Kotlin Multiplatform.

Reply:
Yes, main reason is OOB-ness. I don't want to add something to the runtime that drags in a bunch of transitive dependencies. For that I would prefer the separate module. But I see nothing wrong with providing a basic HttpTransport with maybe not the best performance that needs no dependencies and is this simple as part of the "core" library.

Reply:
Fair point re: "OOB-ness".  My concern definitely is around the default easy path.  Nothing's easier, dependency-wise, than what's built in.  On the other hand, this is an Android library first and foremost, and on Android OkHttp is basically the standard anymore.  I'm OK with leaving it out for now.

As far as expect/actual, this is the mechanism that lets us have interfaces in common code but have platform-specific implementations.  In the eventual iOS release, we'll definitely want to use an `HttpTransport`.  Having an `expect class HttpTransport` in common code is a prerequisite.

Reply:
I was trying to understand the benefit of the `expect` over a plain interface for this usecase. We can already have different implementations of the http transport, they would all just implement `Transport`.

So if I understand correctly, `expect` just tells you "whatever platform this is ported to also needs to implement this thing", which to me is mainly useful for the "common" library to be able to call platform specific code without having the user pass an implementation to some constructor explicitly (there is always going to be exactly one implementation, no more, no less).
To me, plain old interface still sounds totally fitting to me for this use case. Consider the case that we add an `OkHttpClient` in a separate library as well. Would that then work with the `expect`? Or would this now be a "plain" implementation of `Transport`? If yes, why treat both implementations differently?

Please excuse all the text, I am mainly trying to wrap my head around Kotlin Multiplatform.

Reply:
You're right that we _could_ just have platform-specific types that implement `Transport`.  The thing I think lacks there is that we won't have a single `HttpTransport`.  That seems like a miss to me.

Reply:
I never addressed you question about the benefit(s) of `expect class` over an interface.  The difference is that an `expect class` is _concrete_ - one can instantiate one as-is, without needing to consider what other type might implement the interface.